File: dart/fibonacci/lib/fibonacci.dart
import 'dart:io';
import 'dart:core';
import 'package:intl/intl.dart';

/// Fibonacci Series Calculator
///
/// The Fibonacci Series Calculator is a Dart program that calculates
/// and prints the Fibonacci series up to the nth term with arbitrary
/// precision. It includes features such as error handling for input
/// validation, handling special cases where numbers don't end in "th"
/// (e.g. 1, 2, 3), displaying the sum of the series, and providing
/// the current date and time.
///
/// License: This program is in the public domain.

void main() {
  stdout.writeln("============Fibonacci Series Calculator============");
  stdout.writeln("This Program was Written Using: Dart");

  // signalHandler();
  dateAndTime();
  getUserInput();

  stdout.writeln("===================================================");
}

/// Function to get user input
void getUserInput() {
  while (true) {
    stdout.write("Enter the value of n (an integer): ");
    String? input = stdin.readLineSync();

    if (input == null) {
      eofHandler();
    }

    String? trimmedInput = input?.trim();

    if (trimmedInput == "") {
      print("Please enter something...");
      continue;
    } else if (trimmedInput == "exit") {
      print("Exiting the program...");
      break;
    }

    int n = validateInput(trimmedInput);

    if (n > 0) {
      fibonacciSeries(n);
      break;
    } else {
      print("Please enter a valid positive integer.");
    }
  }
}

/// Function to calculate the nth Fibonacci number with arbitrary precision.
void fibonacciSeries(int n) {
  BigInt a = BigInt.zero;
  BigInt b = BigInt.one;
  BigInt temp;
  BigInt sum = BigInt.zero;

  print("Fibonacci series up to the $n${getSuffix(n)} term:");

  List<BigInt> series = [];

  for (int i = 0; i <= n; i++) {
    if (n <= 5000) {
      series.add(a);
    } else {
      // Print the series without using array
      stdout.write("$a");
      if (i < n) {
        stdout.write(", ");
      }
    }

    temp = a;
    a = b;
    b = b + temp;
    sum = sum + temp; // Calculate the sum
  }

  if (n <= 5000) {
    stdout.write(series.join(", "));
  }

  print("\n");
  print("Sum of the Fibonacci series: $sum");
}

/// Handle special cases where numbers don't end in "th"
String getSuffix(int n) {
  if (n % 10 == 1 && n % 100 != 11) {
    return "st";
  } else if (n % 10 == 2 && n % 100 != 12) {
    return "nd";
  } else if (n % 10 == 3 && n % 100 != 13) {
    return "rd";
  } else {
    return "th";
  }
}

/// Display the current date and time
void dateAndTime() {
  var currentDate = DateTime.now();

  String layout = 'MMMM dd, yyyy - HH:mm:ss';
  var formatter = DateFormat(layout);
  String formattedDate = formatter.format(currentDate);

  print("Date and Time: $formattedDate");
}

/// Functions to handle user input and errors
int validateInput(String? input) {
  try {
    int n = int.parse(input ?? '');
    if (n > 0) {
      return n;
    }
  } catch (e) {
    return -1;
  }
  return -1;
}

void eofHandler() {
  print("");
  print("End of File encountered.. Stopping...");
  print("===================================================");
  exit(0);
}

void interruptHandler() {
  print("");
  print("Interrupt received.. Stopping...");
  print("===================================================");
  exit(0);
}

void signalHandler() {
  // Not Working as Intended
  ProcessSignal.sigint.watch().listen((signal) => interruptHandler());
}

File: php/fibonacci/fibonacci.php
<?php

/**
 * Fibonacci Series Calculator
 *
 * The Fibonacci Series Calculator is a PHP program that calculates
 * and prints the Fibonacci Series up to the nth term with arbitrary
 * precision. It includes features such as error handling for input
 * validation, handling special cases where numbers don't end in "th"
 * (e.g. 1, 2, 3), displaying the sum of the series, and providing the
 * current date and time.
 *
 * License: This program is in the public domain.
 */

function main() {
  echo "============Fibonacci Series Calculator============" . PHP_EOL;
  echo "This Program was Written Using: PHP" . PHP_EOL;

  signalHandler();
  dateAndTime();
  getUserInput();

  echo "===================================================" . PHP_EOL;
}

// Function to Get the User Input
function getUserInput() {
  while (true) {
    $input = readline("Enter the value of n (an integer): ");
    if ($input === false) {
      eofHandler();
    }

    $trimmedInput = trim($input);

    if ($trimmedInput === "") {
      echo "Please enter something..." . PHP_EOL;
      continue;
    } elseif ($trimmedInput === "exit") {
      echo "Exiting the program..." . PHP_EOL;
      break;
    }

    $n = validateInput($trimmedInput);

    if ($n !== false) {
      fibonacciSeries($n);
      break;
    } else {
      echo "Please enter a valid positive integer." . PHP_EOL;
    }
  }
}

// Calculates and prints the Fibonacci Series up to the nth term
function fibonacciSeries($n) {
  bcscale(0);

  $a = '0'; // Use strings to ensure precision
  $b = '1';
  $temp = '0';
  $sum = '0';

  echo "Fibonacci Series up to the " . $n . getSuffix($n) . " term:" . PHP_EOL;

  $series = [];

  for ($i = 1; $i <= $n + 1; $i++) {
    if ($n <= 5000) {
      $series[] = $a;
    } else {
      // Print the series without using array
      echo $a;
      echo "" . (($i <= $n) ? ", " : "\n");
    }

    $temp = $a;
    $a = $b;
    $b = bcadd($temp, $b);
    $sum = bcadd($sum, $temp); // Calculate the sum
  }

  if ($n <= 5000) {
    echo implode(", ", $series) . PHP_EOL;
  }

  echo PHP_EOL;
  echo "Sum of the Fibonacci Series: " . $sum . PHP_EOL;
}

// Handle special cases where numbers don't end in "th"
function getSuffix($n) {
  if ($n % 10 == 1 && $n % 100 != 11) {
    return "st";
  } elseif ($n % 10 == 2 && $n % 100 != 12) {
    return "nd";
  } elseif ($n % 10 == 3 && $n % 100 != 13) {
    return "rd";
  } else {
    return "th";
  }
}

// Display the current date and time
function dateAndTime() {
  $current_date = new DateTime();

  $layout = "F d, Y - H:i:s";
  $formatted_date = $current_date->format($layout);

  echo "Date and Time: " . $formatted_date . PHP_EOL;
}


// Functions to handle user input and error
function validateInput($input) {
  if (ctype_digit($input)) {
    $n = intval($input);
    return ($n > 0) ? $n : false;
  }

  return false;
}

function eofHandler() {
  echo "" . PHP_EOL;
  echo "End of File encountered.. Exiting..." . PHP_EOL;
  echo "===================================================" . PHP_EOL;
  exit;
}

function interruptHandler($signal) {
  echo "" . PHP_EOL;
  echo "Interrupt received.. Exiting..." . PHP_EOL;
  echo "===================================================" . PHP_EOL;
  exit;
}

function signalHandler() {
  pcntl_signal(SIGINT, "interruptHandler");
}

main();
